database/import.go

```diff
@@ -0,0 +1,413 @@
+	"github.com/bitcoin-sv/pulse/domains"
+	"github.com/bitcoin-sv/pulse/domains/logging"
+	"github.com/bitcoin-sv/pulse/internal/chaincfg/chainhash"
+	"github.com/bitcoin-sv/pulse/repository"
+)
+
+type DbIndex struct {
+	name string
+	sql  string
+}
+
+type SQLitePragmaValues struct {
+	Synchronous int
+	JournalMode string
+	CacheSize   int
+}
+
+const (
+	insertTransactionSize = 100
+	numWorkers            = 5 // Adjust based on performance and resource constraints
+)
+
+func ImportHeaders(db *sqlx.DB, cfg *config.Config, log logging.Logger) error {
+	log.Info("Import headers from file to the database")
+
+	if !fileExistsAndIsReadable(cfg.Db.PreparedDbFilePath) {
+		return fmt.Errorf("file %s does not exist or is not readable", cfg.Db.PreparedDbFilePath)
+	}
+
+	currentDir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	headersRepo := initHeadersRepo(db, cfg)
+
+	dbHeadersCount, _ := headersRepo.GetHeadersCount()
+
+	if dbHeadersCount > 0 {
+		return fmt.Errorf("database already contains %d block headers, the headers table in the database must be empty", dbHeadersCount)
+	}
+
+	compressedHeadersFilePath := filepath.Clean(filepath.Join(currentDir, cfg.Db.PreparedDbFilePath))
+	tmpHeadersFileName := "headers.csv"
+	tmpHeadersFilePath := filepath.Clean(filepath.Join(os.TempDir(), tmpHeadersFileName))
+
+	log.Infof("Decompressing file %s to %s", compressedHeadersFilePath, tmpHeadersFilePath)
+
+	compressedHeadersFile, err := os.Open(compressedHeadersFilePath)
+	if err != nil {
+		return err
+	}
+
+	tmpHeadersFile, err := os.Create(tmpHeadersFilePath)
+	if err != nil {
+		return err
+	}
+
+	if err := gzipDecompressWithBuffer(compressedHeadersFile, tmpHeadersFile); err != nil {
+		return err
+	}
+
+	if err := compressedHeadersFile.Close(); err != nil {
+		return err
+	}
+
+	log.Infof("Decompressed and wrote contents to %s", tmpHeadersFilePath)
+
+	pragmas, err := getSQLitePragmaValues(db)
+	if err != nil {
+		return err
+	}
+
+	if err := modifySQLitePragmas(db); err != nil {
+		return err
+	}
+
+	droppedIndexes, err := removeIndexes(db)
+	if err != nil {
+		return err
+	}
+
+	if err := importHeadersFromFileConcurrent(headersRepo, tmpHeadersFile, log); err != nil {
+		return err
+	}
+
+	if err := tmpHeadersFile.Close(); err != nil {
+		return err
+	}
+
+	if err = restoreIndexes(db, droppedIndexes); err != nil {
+		return err
+	}
+
+	if err = restoreSQLitePragmas(db, *pragmas); err != nil {
+		return err
+	}
+
+	if err := os.Remove(tmpHeadersFilePath); err != nil {
+		return err
+	}
+
+	log.Infof("Deleted temporary file %s", tmpHeadersFilePath)
+
+	return nil
+}
+
+func initHeadersRepo(db *sqlx.DB, cfg *config.Config) *repository.HeaderRepository {
+	lf := logger.DefaultLoggerFactory()
+	headersDb := sql.NewHeadersDb(db, cfg.Db.Type, lf)
+	headersRepo := repository.NewHeadersRepository(headersDb)
+	return headersRepo
+}
+
+func importHeadersFromFileConcurrent(repo repository.Headers, inputFile *os.File, log logging.Logger) error {
```

Comment:
I am having trouble understanding this code:
1. Errors are placed in the `errCh` channel, but they are never read from the channel.
2. There is no indication of how we wait for workers. I think we should and if we do - how is it done?
3. What does `insertTransactionSize` do? My understanding is that it should represent a bundle/batch size. However, from the code, it seems like we use just `numWorkers` numbers of bundles with non-deterministic sizes and we should use `headersCount / bundleSize` numbers of bundles 

It's possible that my knowledge about go and go routines is not complete to understand this so please correct me.

Reply:
@arkadiuszos4chain @dorzepowski I have asked @augustync, the author of this code, to explain this 😄

Reply:
1. Errors in errCh channel:

The errors placed in the errCh channel are actually consumed in the worker function. In the worker, if there is an error during the database insertion (repo.AddMultipleHeadersToDatabase), the error is sent to errCh. This is a form of error handling, where the main goroutine can later check for errors by receiving from errCh. It might be more explicit to log or handle these errors in the main goroutine as well.

2. Waiting for Workers:

The function uses a combination of channels (doneCh) and goroutine closure to wait for the workers to complete. The defer statement in the worker function sends a signal to doneCh when the worker completes. The main goroutine waits for each worker to finish by receiving from doneCh in a loop (for i := 0; i < numWorkers; i++). This ensures that the main goroutine doesn't proceed beyond the loop until all workers have completed.

3. insertTransactionSize and Batching:

insertTransactionSize represents the number of records to process in each iteration of the outer loop. It acts as a batch size for reading and processing records from the CSV file. The inner loop (for i := 0; i < insertTransactionSize; i++) processes records in batches, and the size of each batch is determined by this variable.
The idea is to efficiently read and process records in chunks, reducing the frequency of database insertions and potentially improving performance. The main goroutine sends blocks to workers in these batches. The use of insertTransactionSize doesn't necessarily imply that each worker gets a bundle; rather, it's about reading records from the file in batches.

Reply:
I was mostly interested in p. 1.
I mean, although we worker could put an error on a channel, this channel isn't read at all. And IMHO actually having this error could lead to inconsistent state of DB, because we could think that we insert all the headers, but actually we didn't because there were some errors and for example 1k of headers weren't inserted.

Reply:
yes that make sense .. that why we can log or handle it better :) feel free to experiment